### PR TITLE
add config tutorial

### PR DESCRIPTION
**Pull request by L1aoXingyu**
给 config 里面的参数写了详细的注释，完成了 config tutorial 相关的文档。

### CODE REVIEW

File: configs/common/train.py [hunk 1]
```diff
@@ -4,64 +4,118 @@
...
 # fmt: off
 train = dict(
+
+    # Directory where output files are written
     output_dir="./output",
 
-    # Batch size and gradient accumulation
+    # `train_micro_batch_size` is number of images per batch on each GPU.
+    # train_mini_batch_size = train_micro_batch_size * num_accumulation_steps.
+    # This is also the number of training images per step (i.e. per iteration).
```
Comment (CPFLAME) [inline configs/common/train.py:13]: 这里用`images` 感觉不太准确, 可以换成`samples`, 或者更合适的

File: configs/common/train.py [hunk 2]
```diff
@@ -4,64 +4,118 @@
...
+    # Total 64 samples will be trained per iteration across all GPUs.
+
+    # global_batch_size = micro_batch_size  * num_grad_acc * data_parallel_groups
     train_micro_batch_size=32,
-    test_micro_batch_size=32,
     global_batch_size=None,
     num_accumulation_steps=None,
-    start_iter=0,
+
+    # The total training iteration
```
Comment (Ldpe2G) [inline configs/common/train.py:25]: iteration -> iterations

File: configs/common/train.py [hunk 3]
```diff
@@ -4,64 +4,118 @@
...
     train_iter=10000,
-    train_epoch=0,  # default train epoch is set to 0
-    warmup_ratio=0,  # default warmup ratio is set to 0
+    # The total training epoch, will be scaled to iteration automatically.
+    # We will choose by `max(train_iter, train_epoch * iter_per_epoch)`.
```
Comment (Ldpe2G) [inline configs/common/train.py:28]: epoch -> epochs,
iteration -> iterations

Comment (Ldpe2G) [inline configs/common/train.py:28]: # The total training iterations will be calculated by the formula `max(train_iter, train_epoch * iter_per_epoch)`.
这样如何

File: configs/common/train.py [hunk 4]
```diff
@@ -4,64 +4,118 @@
...
-    test_micro_batch_size=32,
     global_batch_size=None,
     num_accumulation_steps=None,
-    start_iter=0,
+
+    # The total training iteration
     train_iter=10000,
-    train_epoch=0,  # default train epoch is set to 0
-    warmup_ratio=0,  # default warmup ratio is set to 0
+    # The total training epoch, will be scaled to iteration automatically.
```
Comment (Ldpe2G) [inline configs/common/train.py:27]: epoch -> epochs
iteration -> iterations

File: configs/common/train.py [hunk 5]
```diff
@@ -4,64 +4,118 @@
...
+    # The start iteration, usually needn't set it manually.
+    # It can be computed automatically when resuming training.
+    start_iter=0,
 
-    # Performance related
-    amp=dict(enabled=False),  # options for Automatic Mixed Precision
-    activation_checkpoint=dict(enabled=False),  # options for recompute gradient
-    # NCCL fusion threshold megabytes, set to 0 to compatible with previous version of OneFlow
+    # Enable automatic mixed precision for training
+    # Note that this does not change model's inference behavior.
```
Comment (Ldpe2G) [inline configs/common/train.py:42]: # Enable automatic mixed precision for training which won't change model's inference behavior.
直接合并成一句？

File: configs/common/train.py [hunk 6]
```diff
@@ -4,64 +4,118 @@
...
+    amp=dict(enabled=False),
+
+    # Enable activation checkpointing to allow for training
+    # with larger models, sequences, and batch sizes.
+    # Checkpoint the input activations of each transformer layers by default.
```
Comment (Ldpe2G) [inline configs/common/train.py:47]: 默认 checkpoint 每个 transformer 层的输入是什么意思

File: configs/common/train.py [hunk 7]
```diff
@@ -4,64 +4,118 @@
...
+    # Enable ZeRO Optimization to allow for training with larger models.
+    # This optimization will reduce optimizer stages memory consumption
+    # as described in ZeRO https://arxiv.org/abs/1910.02054.
     zero_optimization=dict(
         enabled=False,
         stage=1,
     ),
 
-    checkpointer=dict(period=5000, max_to_keep=100),  # options for PeriodicCheckpointer
+    # Save a checkpoint after every this number of iterations,
```
Comment (Ldpe2G) [inline configs/common/train.py:66]: Save a model checkpoint ？ 明确一下是保存模型的，不然容易和上面的 `activation_checkpoint` 混淆

File: configs/common/train.py [hunk 8]
```diff
@@ -4,64 +4,118 @@
...
+
+    # Options for evaluation
+
+    # `test_micro_batch_size` is number of samples per batch on each GPU for testing.
+    # If we use 8 GPUs for data parallel groups and `test_micro_batch_size = 2`, then
+    # total 16 samples will be used per iteration across all GPUs.
+    test_micro_batch_size=32,
 
-    # options for evaluation
+    # Enabled evaluation during training every `eval_period` number of iterations.
```
Comment (Ldpe2G) [inline configs/common/train.py:77]: \# Enable evaluation during training, after every `eval_period` number of training iterations will perform the evaluation process.

这样如何

File: configs/common/train.py [hunk 9]
```diff
@@ -4,64 +4,118 @@
...
         eval_period=5000,
-        eval_iter=1e9,  # running steps for evaluation
+        eval_iter=1e9,  # running steps for validation/test
+
+        # Metrics to be used for best model checkpoint.
         eval_metric="Acc@1",
         eval_mode="max",
     ),
 
+    # Path to a checkpoint file to be loaded to the model.
```
Comment (Ldpe2G) [inline configs/common/train.py:92]: \# Path to a checkpoint file for resume training or evaluation ?
是要表达这个意思吗

File: docs/source/tutorials/Config.md [hunk 10]
```diff
@@ -0,0 +1,357 @@
+# Lazy Configs
+
+We find the traditional yacs-based config system or python argparse command-line options cannot offer enough flexibility for new project development. We just borrowed the [lazy config system](https://detectron2.readthedocs.io/en/latest/tutorials/lazyconfigs.html) from detectron2 as an alternative, non-intrusive config system for LiBai.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Config.md:3]: We found that the traditional yacs-based config system or python argparse command-line options suffer from not providing enough flexibility for the development of new project. So we borrow the [lazy config system](https://detectron2.readthedocs.io/en/latest/tutorials/lazyconfigs.html) design from detectron2 which forms the non-intrusive config system for LiBai.

这样如何

File: docs/source/tutorials/Config.md [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+
+You can read the [d2 tutorial](https://detectron2.readthedocs.io/en/latest/tutorials/lazyconfigs.html) for the syntax and basic usage of lazy config. Here we will show you some example usage in LiBai.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Config.md:5]: You can refer to the [d2 tutorial](https://detectron2.readthedocs.io/en/latest/tutorials/lazyconfigs.html) for more details of the syntax and basic usage of lazy config. We will show you some example usages in LiBai in the following secion.

这样如何

File: docs/source/tutorials/Config.md [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+
+## Configs in LiBai
+
+In LiBai, we define a standard set of config namespace for later use. This set of namespace must be kept if you want to use complete training and evaluation process of LiBai.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Config.md:9]: In LiBai, we can define a standard set of config namespaces for later usage. This set of namespaces must be kept if you want to perform the complete training and evaluation process of LiBai. 
这样如何

File: docs/source/tutorials/Config.md [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+
+We find the traditional yacs-based config system or python argparse command-line options cannot offer enough flexibility for new project development. We just borrowed the [lazy config system](https://detectron2.readthedocs.io/en/latest/tutorials/lazyconfigs.html) from detectron2 as an alternative, non-intrusive config system for LiBai.
+
+You can read the [d2 tutorial](https://detectron2.readthedocs.io/en/latest/tutorials/lazyconfigs.html) for the syntax and basic usage of lazy config. Here we will show you some example usage in LiBai.
+
+## Configs in LiBai
+
+In LiBai, we define a standard set of config namespace for later use. This set of namespace must be kept if you want to use complete training and evaluation process of LiBai. 
+
+In summary, this namespace is `model, graph, train, optim, dataloader, tokenization(optional)`, and we will introduce it in detail below.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Config.md:11]: In summary, this set of config namespaces are `model, graph, train, optim, dataloader, tokenization(optional)`, and we will introduce them in the following in detail.

这样如何

File: docs/source/tutorials/Config.md [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+
+### model
+
+This is the config for model definition. You can see some examples in `configs/common/models`.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Config.md:15]: The configuration for model definition. You can refer to `configs/common/models` for more examples.

这样好些

File: docs/source/tutorials/Config.md [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+model.hidden_layers = 12 # change hidden layers
+```
+
+After you define the model config in a python file, you can `import` it in the global scope of the config file. Note that you need to rename it as `model` regardless of the name used in model config.
+
+You can access and change all keys in the model config after import.
+
+### graph
+
+This is the config for `nn.Graph` mode. You can learn more information about the static graph mode in official [nn.Graph docs](https://docs.oneflow.org/master/basics/08_nn_graph.html).
```
Comment (rentainhe) [inline docs/source/tutorials/Config.md:58]: 感觉用`learn from`会好些

File: docs/source/tutorials/Config.md [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+
+LiBai has already defined a `GraphBase` class for almost all models use. You can simply turn on this option converting eager mode to graph mode. 
+
+The graph config can be found in `configs/common/models/graph.py`, and two useful options are shown as follows:
```
Comment (rentainhe) [inline docs/source/tutorials/Config.md:62]: 这儿是不是给个链接比较好 can be found in [graph.py]()

Comment (rentainhe) [reply docs/source/tutorials/Config.md:62]: 我不确定用链接好一些还是直接给路径好一些，因为安装完libai后这个路径就在`libai/config/configs`下了，是不是也可以说一下可以用`get_config`这个函数得到, `get_config("common/models/graph.py")`

File: configs/common/train.py (hunk 6 above)
Comment (L1aoXingyu) [reply configs/common/train.py:47]: 这个是指重计算的时候，保存每个 TransformerLayer 的输入，算梯度的时候，中间的 activation 重新算

Comment (Ldpe2G) [reply configs/common/train.py:47]: 好的，是指  `enabled=True` 开启之后，默认的策略是吧，那可以在开头加一下
 If enabled, Checkpoint the input .....

File: docs/source/index.rst [hunk 17]
```diff
@@ -14,6 +14,7 @@ Welcome to libai's documentation!
    tutorials/Installation.md
    tutorials/Getting Started.md
    tutorials/Train and Eval Command Line.md
+   tutorials/Config.md
```
Comment (rentainhe) [inline docs/source/index.rst:17]: 感觉`Config System`更贴切(optional)

File: docs/source/tutorials/Config.md [hunk 18]
```diff
@@ -0,0 +1,359 @@
...
+    weights=[1.0],
+    num_workers=4,
+)
+
+# my_config.py:
+dataloader.train.dataset[0].max_seq_length = 256
+dataloader.train.num_workers = 2
+```
+
+LiBai provides two functions `build_nlp_train_val_test_loader` and `build_image_train_loader` that create a default train data loader from a give config. It takes the list of `dataset_class`(e.g., `BertDataset`) and combine them with a `ConcatDataset`.
```
Comment (rentainhe) [inline docs/source/tutorials/Config.md:286]: combine them using `flow.utils.data.dataset.ConcatDataset`感觉更合适 (optional)

**Review by rentainhe — APPROVED** [review]

File: docs/source/index.rst [hunk 19]
```diff
@@ -14,6 +14,7 @@ Welcome to libai's documentation!
    tutorials/Installation.md
    tutorials/Getting Started.md
    tutorials/Train and Eval Command Line.md
+   tutorials/Config_System.md
```
Comment (rentainhe) [inline docs/source/index.rst:17]: 似乎不需要加下划线2333

Comment (L1aoXingyu) [reply docs/source/index.rst:17]: 这个只是文件的名字，文档里面的名字是 config_system.md 里面第一个 # 的内容

Comment (rentainhe) [reply docs/source/index.rst:17]: 嗷嗷，get到了，我一直以为这里会对应= =

**Review by rentainhe — APPROVED** [review]

File: docs/source/tutorials/Config_System.md [hunk 20]
```diff
@@ -0,0 +1,359 @@
+# Config System
+
+We found the traditional yacs-based config system or python argparse command-line options suffer from providing enough flexibility for the development of new project. So we borrowed the [lazy config system](https://detectron2.readthedocs.io/en/latest/tutorials/lazyconfigs.html) design from detectron2 which forms the non-intrusive config system for LiBai.
+
+You can refer to the [d2 tutorial](https://detectron2.readthedocs.io/en/latest/tutorials/lazyconfigs.html) for more details of the syntax and basic usage of lazy config. We will show you some example usage in LiBai in the following section.
+
+## Configs in LiBai
+
+In LiBai, we define a standard set of config namespaces for later use. This set of namespaces must be kept if you want to perform the complete training and evaluation process of LiBai.
```
Comment (CPFLAME) [inline docs/source/tutorials/Config_System.md:9]: `set` 是不是应该叫 `setting`

Comment (L1aoXingyu) [reply docs/source/tutorials/Config_System.md:9]: a set of 是一组的意思

Comment (CPFLAME) [reply docs/source/tutorials/Config_System.md:9]: 可以的

File: docs/source/tutorials/Config_System.md [hunk 21]
```diff
@@ -0,0 +1,359 @@
...
+    weights=[1.0],
+    num_workers=4,
+)
+
+# my_config.py:
+dataloader.train.dataset[0].max_seq_length = 256
+dataloader.train.num_workers = 2
+```
+
+LiBai provides two functions `build_nlp_train_val_test_loader` and `build_image_train_loader` that create a default train data loader from a give config. It takes the list of `dataset_class`(e.g., `BertDataset`) and combine them using `flow.utils.data.dataset.ConcatDataset`.
```
Comment (kaijieshi7) [inline docs/source/tutorials/Config_System.md:286]: It takes and combines ... ?

File: docs/source/tutorials/Config_System.md [hunk 22]
```diff
@@ -0,0 +1,359 @@
...
+    do_chinese_wwm=True,
+)
+tokenization.append_eod = False
+tokenization.make_vocab_size_divisible_by = 128
+
+# my_config.py:
+tokenization.tokenizer.do_lower_case = False
+```
+
+Tokenization config must contain the a tokenizer(e.g., `BertTokenizer`). `append_eod` and `make_vocab_size_divisible_by` are not necessary.
```
Comment (kaijieshi7) [inline docs/source/tutorials/Config_System.md:316]: the a ?

**Pull request merged**